[AMDGPU] Implement llvm.lround intrinsic lowering.

sgundapa opened the pull request:
This patch enables the  target-independent lowering of llvm.lround via GlobalISel. For SelectionDAG, the instrinsic is custom lowered for AMDGPU. In order to support vector floating point input for llvm.lround, this patch extends the target independent APIs and provide support for scalarizing. pr98950 is needed to let verifier allow vector floating point types

sgundapa committed:
[AMDGPU] Implement llvm.lround intrinsic lowering.

This patch enables the  target-independent lowering of llvm.lround via
GlobalISel. For SelectionDAG, the instrinsic is custom lowered for AMDGPU.
In order to support vector floating point input for llvm.lround, this patch
extends the target independent APIs and provide support for scalarizing.
pr98950 is needed to let verifier allow vector floating point types

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -2522,6 +2526,13 @@ SDValue AMDGPUTargetLowering::LowerFROUND(SDValue Op, SelectionDAG &DAG) const {
   return DAG.getNode(ISD::FADD, SL, VT, T, SignedOffset);
 }
 
+SDValue AMDGPUTargetLowering::LowerLROUND(SDValue Op, SelectionDAG &DAG) const {
```
This should just use the default handling which should be the same

arsenm reviewed (commented):
There's still too much going on here. We don't really need vector support other than for consistency, we're just going to scalarize it anyway. The vector support should be done completely separately, after the scalar case is handled.

The AMDGPU lowering should not require custom lowering, the default expansion path can be the same (although you might need to fix the default action to be libcall instead of expand).

The TTI changes are also separate

sgundapa committed:
Implement llvm.lround lowering for GlobalISel and SelectionDAG.

Also, removed handling of vector types. Support will be pushed in a later patch

sgundapa committed:
Update X64 rule for llvm.lround

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/lround.ll:
```diff
@@ -0,0 +1,551 @@
+; GFX11-NEXT:    v_sub_co_ci_u32_e32 v1, vcc_lo, v1, v3, vcc_lo
+; GFX11-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %0 = tail call i64 @llvm.llround.i64.f32(float %arg)
```
Use named values in tests

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/lround.ll:
```diff
@@ -0,0 +1,551 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx900 < %s | FileCheck -check-prefixes=GCN,GFX9 %s
```
Just add global isel run lines to the test in the AMDGPU directory instead of duplicating the whole file

arsenm reviewed (commented):
lgtm with the test merged

sgundapa committed:
Address comments

arsenm commented on llvm/test/CodeGen/AMDGPU/lround.ll:
```diff
@@ -0,0 +1,1036 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --extra_scrub --version 5
+; RUN: llc -global-isel=0 -mtriple=amdgcn -mcpu=gfx900 < %s | FileCheck -check-prefixes=GFX9-SDAG %s
+; RUN: llc -global-isel=1 -mtriple=amdgcn -mcpu=gfx900 < %s | FileCheck -check-prefixes=GFX9-GISEL %s
```
Should add a common GFXN check prefix for all of these

sgundapa replied:
I have tried adding it and there are no common lines
